pkg/proxy/iptables/proxier_test.go

```diff
@@ -5291,7 +5291,7 @@ func TestProxierDeleteNodePortStaleUDP(t *testing.T) {
 			eps.Endpoints = []discovery.Endpoint{{
 				Addresses: []string{epIP},
 				Conditions: discovery.EndpointConditions{
-					Ready: pointer.Bool(true),
+					Serving: pointer.Bool(true),
```

Comment:
We should fix this everywhere. `BaseEndpointInfo.Ready` should just go away, and `BaseEndpointInfo.IsReady()` should return `info.Serving && !info.Terminating`.

Reply:
> Do you mean to completely remove the property `Ready` from `BaseEndpointInfo` structure itself ?

Yes. It's redundant; `info.Ready` is guaranteed to always be the same as `info.Serving && !info.Terminating`, because that is true of the corresponding EndpointSlice fields that they're set from.

Reply:
setting `Terminating: false` for this one.

Reply:
same for this, https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/proxy/ipvs/proxier_test.go#L4577-L4581

Reply:
Yeah, `Ready: false, Serving: true, Terminating: false` is an invalid combination of conditions. The test cases are bad.

> setting `Terminating: true` for this one.

(that might make it redundant with an already-existing test case)

Reply:
we can only ignore Ready if Serving is exactly the same, if we have some corner case that is not covered as reported here https://github.com/kubernetes/kubernetes/issues/119773 we can not modify an existing behavior ... 

There was an incoherence when setting the `publishNotReady` field, check if this is the problem here https://github.com/kubernetes/kubernetes/issues/116285

Reply:
endpoint slice controller also considers `PublishNotReadyAddresses` in determining the Ready state. 
https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/staging/src/k8s.io/endpointslice/utils.go#L36-L42

Reply:
@danwinship should I remove this commit and use **EndpointConditions.Ready** condition only ?

Reply:
The API docs for `EndpointConditions` say

>     // serving is identical to ready except that it is set regardless of the
>     // terminating state of endpoints.

And the docs for `publishNotReadyAddresses` say

>     // publishNotReadyAddresses indicates that any agent which deals with endpoints for this
>     // Service should disregard any indications of ready/not-ready.

Which means that given the current state of the endpointslice controller, kube-proxy does not handle terminating endpoints correctly for Services with `publishNotReadyAddresses` set, because it looks at `serving`, not `ready`.

I would say that the quoted endpointslice utils code is incorrect; it ought to be overriding `serving` based on `publishNotReadyAddresses`, _not_ overriding `ready` based on that, and it ought to always be setting `ready` to `serving && !terminating`.

However, that is getting a bit far afield from the rest of this PR, so I guess just leave the maybe-redundant-maybe-incorrect separate handling of Ready and Serving in `pkg/proxy` for now.

Reply:
the api doc was fixed in https://github.com/kubernetes/kubernetes/pull/116743/files

Reply:
>  always be setting ready to serving && !terminating.

+1

Reply:
Agree, since previously we just removed terminating endpoints entirely, right?